Loading with FITS "wildcard" indexes

ejeschke opened the pull request:
- added a new loading API in Control module to use a continuation based approach that is less tightly synchronized and more flexible than the older load_file() method (which is kept, for the time being)
- added a Downloads plugin that handles URI downloads
- FITS openers now are expanded to be able to load wildcarded indexes (e.g. [SCI,*], [*,2], etc.)
- modified the FileSelection dialogs so that they can optionally make a single callback with all selected files.  The old method of making a callback for each filepath is kept for now (but should be deprecated!)

EDIT:
* Fix #410 
* Fix #559 
* Close #734 
* Address part of #311

ejeschke commented:
@pllim, alternative to #734.  Sorry it took so long, I had to wrestle with this one to find the right approach.  LMK if it provides the functionality you were seeking.  *This still needs some polish before it is ready to merge*.  But I think it is ready for testing.

ejeschke commented:
BTW, you can check out the new `Downloads` plugin by opening it up, and then dragging some URIs [from here](https://fits.gsfc.nasa.gov/fits_samples.html) into a channel or the `Thumbs` pane.

ejeschke commented:
@pllim, added your fixes comments and documentation on the new plugin.  I am done polishing it.

ejeschke commented:
Rebased.

ejeschke commented on ginga/rv/Control.py:
```diff
@@ -723,6 +697,181 @@ def load_file(self, filepath, chname=None, wait=True,
+            The parameter is the local filepath after download, plus
+            any "index" understood by the loader.
+
+        If the file is already present (e.g. a file:// URI) then this
```
Done

ejeschke commented on ginga/rv/main.py:
```diff
@@ -622,15 +624,21 @@ def main(self, options, args):
+        else:
+            i = 0
+            for imgfile in expanded_args:
+                if i < len(channels):
```
Done

ejeschke commented on ginga/rv/plugins/Downloads.py:
```diff
@@ -0,0 +1,203 @@
+        try:
+            # Try to download the URL.  We press our generic URL server
+            # into use as a generic file downloader.
+            dl = catalog.URLServer(self.logger, "downloader", "dl",
```
Yes, or maybe deprecate it in favor of the Python `requests` module! (or something similar)

ejeschke replied:
For a future PR...

pllim replied:
Okay, deferred to #759.

ejeschke commented on ginga/rv/plugins/SAMP.py:
```diff
@@ -332,7 +332,8 @@ def display_fitsfile(self, chname, fitspath, dowait):
         then wait for the file to be loaded before returning (synchronous).
         """
-        self.fv.load_file(fitspath, chname=chname, wait=dowait)
+        # TEMP: dowait ignored
```
Done

ejeschke commented on ginga/rv/plugins/TVMark.py:
```diff
@@ -512,6 +513,16 @@ def load_file(self, filename):
 
         self.redo()
 
+    def load_files(self, filenames):
```
Fixed.  This also should fix the bug with `TVMark` that you showed above. Would you mind checking?

pllim commented on ginga/rv/plugins/Downloads.py:
```diff
@@ -0,0 +1,205 @@
+plugin from under the "Util" category.
+
+If you want to initiate a download, simply drag a URI into a channel
+image viewer or the `Thumbs` pane.
```
```suggestion
image viewer or the ``Thumbs`` pane.
```

pllim commented on ginga/rv/plugins/Downloads.py:
```diff
@@ -0,0 +1,205 @@
+**Settings**
+
+The ``auto_clear_download`` option, if set to `True`, will cause a download
+entry to be automatically deleted from the pane when the download completes.
```
```suggestion
entry to be automatically deleted from the pane when the download completes.
It does not remove any downloaded file(s).
```

pllim commented on ginga/rv/plugins/SAMP.py:
```diff
@@ -329,10 +329,12 @@ def __init__(self, fv, logger):
     def display_fitsfile(self, chname, fitspath, dowait):
-        """Load (fitspath) into channel (chname).  If (dowait) is True
-        then wait for the file to be loaded before returning (synchronous).
+        """Load (`fitspath`) into channel (`chname`).
```
```suggestion
        """Load (``fitspath``) into channel (``chname``).
```

pllim commented on ginga/rv/plugins/SAMP.py:
```diff
@@ -329,10 +329,12 @@ def __init__(self, fv, logger):
-        then wait for the file to be loaded before returning (synchronous).
+        """Load (`fitspath`) into channel (`chname`).
+
+        (The parameter `dowait` is currently ignored.)
```
```suggestion
        (The parameter ``dowait`` is currently ignored.)
```

pllim commented on ginga/rv/plugins/Downloads.py:
```diff
@@ -0,0 +1,203 @@
+        def _dl_indicator(count, blksize, totalsize):
+            # calculate elapsed time label
+            cur_time = time.time()
+            elapsed = cur_time - track.time_start
```
Is this not what you are trying to do?

```python
>>> from datetime import datetime
>>> d1 = datetime.now()
>>> # Wait a second or two
>>> d2 = datetime.now()
>>> dt = d2 - d1
>>> str(dt)
'0:00:04.000012'
```

Ref: https://stackoverflow.com/questions/538666/python-format-timedelta-to-string

ejeschke replied:
Why yes it is!  (didn't think to check the `str` of it).

ejeschke replied:
Although it looks like I might have to tweak the output a bit to get hours padded and to remove the fractional seconds.  I think I might leave this for a later fix.

pllim replied:
Weird that it is so hard to customize the formatting but maybe a suggestion from https://stackoverflow.com/questions/18470627/how-do-i-remove-the-microseconds-from-a-timedelta-object would work:

```python
str(dt).split(".")[0]
```

ejeschke replied:
> str(dt).split(".")[0]

Still need to pad hours, if we want a fixed size display.

Since I'm already subtracting two Python `time.time()` objects I am getting the result in seconds.  Not sure that the machinations that are necessary to force the `timedelta` object into the right format is prettier or more understandable than just dividing seconds by the necessary constants.  Plus, if the maintainers of `datetime` decide to change the `str` representation of `timedelta` objects then that way breaks, whereas the current scheme doesn't.

pllim replied:
That's true. I didn't realize formatting `timedelta` can get painful if its output format has to be customized. I thought for sure it is implemented already. Yeah, fine, let's stick with the current implementation then. Thanks for humoring me.

pllim commented on ginga/rv/plugins/Downloads.py:
```diff
@@ -0,0 +1,203 @@
+
+**Usage**
+
+Open this plugin to monitor the progress of URI downloads.  Start it
```
Okay, deferred to #760

pllim commented:
Also, I still can't figure out where the data file gets downloaded to when I drag-and-drop into `Thumbs`. Can you please clarify it on the `Downloads` docstring as well? Thanks!

ejeschke commented:
I've updated it to check the setting `download_folder` that you can put in `general.cfg`.  If that is None (or no setting) then it defaults to a directory in the platform temp area as defined by Python's 'tempfile' module.

ejeschke commented:
@pllim, see if you are ok with the bit I added about the download area in the docstring for `Downloads` (I also added it to the example for `general.cfg`).

ejeschke commented:
Good reviews, @pllim!  Merging.
